### ISSUE TITLE
Data leakage in model training

### ISSUE THREAD

**Issue by n33lkanth** [sharathadavanne/multichannel-sed-crnn#11]
Hi @sharathadavanne,
Thank you very much for sharing the code as it helps a lot of ML aspirants like me.
As I was going through the training pipeline, I noticed that :

**1.** You are using the same data as validation and test data while training and testing the model. Isn't this a leakage of information to the model?  Shouldn't we use say 10% of data from training data as validation data and keep test data unseen for the model? **[Reference](https://machinelearningmastery.com/difference-test-validation-datasets/)**

**2.** As per my understanding participants are provided two test datasets i.e., one during the development phase with known targets and the other dataset as an evaluation dataset without known targets. Is my understanding correct? If not then kindly help me to understand this.

**3.** The DCASE 2017 Task 3 challenge [page](http://dcase.community/challenge2017/task-sound-event-detection-in-real-life-audio) says that metrics should be calculated only after finishing the training on all 4 folds. But, In this solution, I can see that metric is being calculated in every fold and later it's being averaged.

**4.** In the README file of this repo you have mentioned that "_The results vary from the original paper, as we are not using the evaluation split here_". I wonder how is it going to affect the Development sub-dataset result if you are not using the evaluation dataset for testing. Even the case of the scores submitted on [channel page](http://dcase.community/challenge2017/task-sound-event-detection-in-real-life-audio-results#Ash_1) for the development dataset(ER: 0.2500 F1: 79.3 %) is very different from the one shown in the README file (ER:  0.60 F1: 0.59/0.57) of this repo.

Kindly share your thought on these points as it will really help me and others to implement follow the training pipeline.

Best Regards

**Comment by sharathadavanne** [comment]
Hi, @n33lkanth thanks for checking the repo. I understand from your questions that you might have confused this repo to be the official DCASE repository. I just want to clarify, this is NOT the official DCASE baseline implementation. The answers to your questions follow,

1. Yes you are right, that is the correct way of evaluation. This repo was written as an example usage of CRNN for SED when such a code and training procedure was not publicly available. I didn't handle any other bells and whistles :) It is my bad that I didn't spend time adding all the scientific approaches. 
2. Yes, you are right. 
3. Yes, you are right, the metrics are not computed as expected in DCASE 2017 task. As I mentioned in 1. the main idea behind this repo was to show the implementation of CRNN and its training procedure. More frankly, I didn't expect the scores to vary much between the way I am computing and the official way of computing, so I ignored the official approach for my ease of implementation and quick release of this repo.
4. Yes, you are right, the development scores in the paper are significantly different from the ones in this repo, because I had a [bug in my metric code while writing the paper](https://github.com/sharathadavanne/sed-crnn/issues/2). That is why you see a significantly large difference in performance between the development and evaluation datasets in the channel page you shared above. The development performance was computed using my buggy code, while the evaluation performance was computed using the official metric code by the organisers (and hence more reliable as a comparison score across DCASE submissions).

Feel free to write to me if you have more doubts. Cheers.

**Comment by n33lkanth** [comment]
Hi @sharathadavanne,
Thank you very much for taking out time and providing detailed feedback.

1. So, as you mentioned that scores in README are from the buggy code. I can see that you [fixed the bug](https://github.com/sharathadavanne/sed-crnn/issues/2) later. Is it Ok to use the fixed code now?

2. If I first train and validate the 4-folds on 90-10 split of training data, save the best model with the lowest error rate in each fold (that means I will have 4 models once all folds are finished), and later during the Test phase load the models from each fold to predict on corresponding test fold (from development dataset) and accumulate the results from each test fold and finally calculate the evaluation metric for development test data. Is this the correct approach?

3. In the case of the evaluation dataset, there are no folds So which model (out of 4 models from different folds) should be used to perform prediction on the Evaluation dataset? Is there a fixed rule or should I go with any model with lowest ER?

4.  _I noticed that in your case you are generating the sequence of length 256 during the preprocessing phase and these sequences are not overlapping. But, in my case, I need to cut the full Mel-spectrogram feature sequence into small segments (of size 512-time frames) with a hop size of 0.8t (0.8*512= 409.6 time frames). So, is it Ok to calculate the no. of time frames in a 1-second segment using the formula you used in this repo for ER & F1 calculation, or do I need to modify it?_

4. As you said that this is not an official baseline code. So, if possible can you please share the code used by you for the challenge as I am finding the official baseline code a bit difficult to understand. It will really help me to implement my model. I chose to go with the training pipeline used in this repo as I find this implementation pretty easy to understand : )

Stay safe!

Best Regards

**Comment by sharathadavanne** [comment]
1. yup the metric code is good to use now.

For the remainder of the questions, I will start with a disclaimer, it has been a while, so my comments may not be 100% accurate.
2. The dataset already gives you four folds of data with pre-defined 'train' and 'evaluate' portions.  You might have to go through the official baseline on how they use this. But if I had to train it the 'scientific way', I would split the 'train' portion of the data into a 90-10% split, use the 90% for training, 10% for deciding when to early stop during the training, and finally test this model on the 'evaluate' portion of the data. This way, post-training four different models, each corresponding to the four-folds, you will have the predicted labels for the 'evaluate' portions of all four-folds. However, since not all participants used the baseline code, the training procedures might vary across submissions and there is no easy way to figure out their respective training procedures. Hence even if you follow a certain training strategy your scores may not be directly comparable to the DCASE submissions. My suggestion here would be a) read the papers with which you want to compare the results with and copy their training procedure (if they have described it). b) An easier strategy would be to implement the existing method yourself or use their repo (like this one), run it on your machine, get the results for it, and then compare it with your approach. However, in this case, report both the numbers mentioned in the reference paper and the numbers you obtained by running their code on your machine. This will be a fair comparison.

3. The organisers don't restrict you on how you use the dev-set for the final evaluation. So for my official submission to the challenge, I remember training the model with 90% of the development-dataset recordings, and early stopping on the remaining 10%. There is no 'correct' early stopping metric, I chose the lowest ER model because the submissions were rated based on the best ER, irrespective of how good or bad the F-score was. 

4. You will have to figure out a way to get the final frame-wise results for a recording from your overlapped frame results. Thereafter it is straightforward to use the metric.

5. Frankly, I used this exact repo for the challenge. Hence I ended up reporting the buggy metric code numbers in the paper. I know its not a fair comparison but refer to point 2, to keep it fair :)

**Comment by n33lkanth** [comment]
@sharathadavanne ,
Thanks a lot for so detailed answers as it's very hard to find such detailed explanation these days 👍 
Your feedback really helped me to get the clarity about the training pipeline, evaluation, metric and Dataset.


Best Regards
